fix #92521: make pulseaudio support optional

thomasa88 opened the pull request:
BUILD_PULSEAUDIO is default ON, but will silently exclude pulseaudio
support if pulseaudio is not found or not supported by the
platform. This is in line with the behavior of BUILD_LAME.

shoogle commented:
Welcome aboard!

Funnily enough, I was just in the middle of patching BUILD_LAME to make it [behave like JACK](https://github.com/thomasa88/MuseScore/blob/92521-optional-pulseaudio/CMakeLists.txt#L300) and fail if it isn't found after being requested.

@lasconic, any thoughts?

thomasa88 commented:
Thanks! One thought on Jack: Does the AUTO value work? I tried getting the BUILD_PULSEAUDIO option to produce AUTO but I only managed to get ON and OFF. 

So I didn't find a (trivial) way for the user to be explicit about Pulseaudio on/off and still allow the automatic behavior if the user is not explicit.

lasconic commented:
Hi and welcome @thomasa88,
I don't really have a preference. I just want to mention that JACK and PulseAudio are audio backend while LAME is only used for MP3 export. Actually, we only need the lame headers, the binary is loaded dynamically at runtime. Also, consistence is good. If you can make all these library search consistent, it's great.

Regarding AUTO, I guess it was working before we moved to using OPTION from a variable. OPTION cannot be AUTO, only ON or OFF. https://cmake.org/cmake/help/v3.0/command/option.html

thomasa88 commented:
There's quite a bunch of build options, but I think I could make the behavior for the audio backends behave in the same way. I played around with CMakeLists.txt and JACK is ON by default and indeed does fail if JACK is not found. Going by this, I guess it would be OK to have Pulseaudio default ON by default and users (well, mostly maintainers I guess), will have to define it to OFF if they don't have Pulseaudio.

My interpretation of the build file is that ALSA is always required on Linux/BSD. But it is possible to disable ALSA in the Musescore settings. However, it does control the HAS_MIDI flag.. but MIDI should work with just e.g. JACK? (Was going to try building without ALSA, but I think I have screwed up my QT environment at the moment.). Any input on this? Should it be an option?

thomasa88 commented:
I cleaned up BUILD_JACK a bit and added a BUILD_ALSA option. I could run Musescore without ALSA without problems - and use my MIDI keyboard. Still a bit unsure if HAS_MIDI should be enabled for other things than ALSA.

Currently ALSA and PulseAudio are default ON and an error is thrown if trying to use them with Apple/Mingw. I'm thinking of just make it a warning, and also update the Makefile.<platform> files to have sane defaults (i.e. both are OFF). Sounds good?

Btw, do I need to mention you for you to get notified of new comments? @lasconic @shoogle (I understand if you just haven't had time to reply - I will be more busy, starting tomorrow).

shoogle commented:
If I understood you correctly:
- All backends optional but ON by default in CMakeLists.txt
- If a backend is not supported by a system (e.g. Windows) then it is overriden as OFF in the system's Makefile (Makefile.mingw)
- If a backend is ON but not found then a warning is printed and the build continues.

Sounds good to me. Just make sure those warnings are nice and obvious! Maybe set them to be of [mode "WARNING"](https://cmake.org/cmake/help/v3.0/command/message.html) rather than "STATUS" or none.

thomasa88 commented:
@shoogle: Yes, but with some differences to the third point. So:

* All backends optional but ON by default in CMakeLists.txt
* If a backend is not supported by a system (e.g. Windows) then it is overriden as OFF in the system's Makefile (Makefile.mingw)
* If a backend is ON and
 * is **not** supported by a system then a warning is printed and the build continues.
 * is supported by a system, and not found, then an error occurs and the build will be aborted.

So on a system where it would work to use a backend (e.g. Pulseaudio on Linux) with Musescore, it will fail if the backend is not found. On the other hand, on a system where it does not work to use the backend, only a warning will be printed (e.g. Pulseaudio on Windows).

The more I think about this, it seems more and more sneaky to just issue a warning, but I'm thinking it will avoid breaking with peoples' existing build setups.

I will see if I can figure out good messages. Will definitely be WARNING.

shoogle commented:
It seems like you're giving yourself a bit more work than is strictly necessary because the system is essentially checked twice:
- Firstly, the system Makefile (e.g. Makefile.mingw) disables unsupported components (e.g PulseAudio).
- But then in CMakeLists.txt you check the system again and disable the unsupported component again.

I know you are doing this in case someone calls CMake directly (without using the Makefile), but there might be a neater way to do it.

How about this inside CMakeLists.txt?
<pre>
# Disable components not supported on Windows
if (NOT MINGW)
    option(BUILD_PULSEAUDIO    "not available on Windows" OFF)
endif (NOT MINGW)

# GENERIC OPTIONS
option(BUILD_PULSEAUDIO    "enable the PulseAudio sound server" ON)
option(AEOLUS        "enable pipe organ synthesizer"      OFF)
option(ZERBERUS      "enable experimental sfz sampler"    ON)
option(OSC           "enable OSC remote control protocol" ON)
...
</pre>

Now you don't need to add anything to the Makefile, you don't need to check the system later on, and any Windows users who are using a CMake GUI will see why PulseAudio is disabled for them.

Notes:
- Options can only be set once, hence why the system-specific code must go first.
- If you want to test this you will need to run "make clean" or delete your CMakeCache.txt between builds.

thomasa88 commented:
Thanks for picking me up out of the rabbit hole. Your alternative sounds much more clean and concise.

What still itches a bit is that compatibility logic will still be duplicated - but it is not too bad. Other (worse?) alternatives I can think off:
1. No sane defaults - Will break the Makefile* files
2. Hide the option when not compatible (IF (NOT MINGW)  option(BUILD_PULSEAUDIO) ENDIF) - Will not block the user from doing -DBUILD_PULSEAUDIO=ON anyway
3. A list of compatible/incompatible options for each platform in the CMake file, which is then used to set default options and when doing the build setup. Will require much more glue logic, but might ultimately avoid duplication. Without knowing much CMake yet, it seems it can add too much complexity.

shoogle commented:
There's no need to check the system twice. I don't think you need to worry about Windows users manually setting -DBUILD_PULSEAUDIO=ON. If they deliberately go to the trouble of manually enabling a component that they have already been warned isn't available then they deserve to have it fail on them ;)

The full implementation would be, what I suggested above, and then:

<pre>
# PulseAudio
if (BUILD_PULSEAUDIO)
    # Look for PulseAudio and, if it is not found, EITHER
            # send an error message and fail the build
     # OR
            # send a warning
            # set  USE_PULSEAUDIO  to  FALSE
            # continue with the build
endif (BUILD_PULSEAUDIO)
</pre>

I'll leave it up to you whether you want it to fail or print a warning. There are pros and cons to both but either option is perfectly justifiable so don't spend too long thinking about it.

It's important that we protect against people being lazy and not setting some options, but I don't think we need to worry about people setting extra options.

thomasa88 commented:
I played around a bit with the options and I think I got everything correct now. One gotcha is that I only have a Linux system to test with, but I have injected APPLE and MINGW and seen that the options seem to react as intended (have not overriden anything else though).

I went with the option to fail if a backend is not found, and it seems it broke the CI..

shoogle commented:
How ironic, well that's what the CI is there for.

I think if you add "libpulse-dev" to the list of packages in ".travis.yml" then it will pass.

shoogle commented:
Looks like it passed. Nice job! I think the code is much neater than before, and I'm a fan of having it fail early if a requested dependency is missing because it should prevent package maintainers from missing stuff out accidentally and the having to re-package. Of course, it's up to @lasconic what gets merged.

lasconic commented:
I can compile this branch on Mac OSX and on Windows.

lasconic commented:
Would you mind squashing the commits together with a commit message *fix #92521: make pulseaudio support optional*? Thank you!
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

thomasa88 commented:
No problems! Just pushed it.
